cargo-concordium/src/build.rs

```diff
@@ -93,15 +101,244 @@ impl SchemaBuildOptions {
+            continue;
+        }
+        let relative_path = file.path().strip_prefix(package_root_path)?;
+        if relative_path == std::path::Path::new("Cargo.lock") {
+            lock_file_found = true;
+        }
+        archived_files.push(relative_path.to_path_buf());
+        tar.append_path_with_name(file.path(), relative_path)?;
+    }
+    anyhow::ensure!(
+        lock_file_found,
+        "Unable to proceed with a verifiable build. A Cargo.lock file must be available and up to \
+         date. Run `cargo check` to generate it."
+    );
+    Ok(TarArchiveData {
+        tar_archive: tar.into_inner()?,
+        archived_files,
+    })
+}
+
+/// Build an archive and return the Wasm source that was built.
+pub fn build_archive(
+    image: &str,
+    tar_contents: &[u8],
+    container_runtime: &str,
+    build_command: &[String],
+) -> anyhow::Result<Vec<u8>> {
+    let artifact_dir =
+        tempfile::tempdir().context("Unable to create temporary build directory.")?;
+    // Construct the mapping of the host's build directory
+    // into the container's artifacts directory. That is where the
+    // wasm module will be published.
+    let mapping = {
+        let mut mapping = artifact_dir.path().as_os_str().to_os_string();
+        mapping.push(":/artifacts:Z");
+        mapping
+    };
+
+    let tar_file_name = "archive.tar";
+
+    std::fs::write(artifact_dir.path().join(tar_file_name), tar_contents)
+        .context("Unable to write archive.")?;
+
+    let container_output_dir = "/b/t/wasm32-unknown-unknown/release";
+    let mut cmd = Command::new(container_runtime);
+
+    cmd.arg("run")
+        .arg("-v")
+        .arg(mapping.as_os_str())
+        .arg("--rm")
+        .args(["--workdir", "/b"])
+        .arg(image)
+        .arg("/run-copy.sh")
+        .arg(format!("/artifacts/{tar_file_name}"))
+        .arg(container_output_dir)
+        .args(build_command);
+
+    let result = cmd
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .output()
+        .context("Could not use cargo build.")?;
+
+    if !result.status.success() {
+        anyhow::bail!("Compilation failed.")
+    }
+
+    let filename = artifact_dir.path().join("out.wasm");
+    let wasm = std::fs::read(filename).context("Unable to read generated Wasm artifact.")?;
+    Ok(wasm)
+}
+
+struct ContainerBuildOutput {
+    /// The output Wasm file containing the unprocessed contract module.
+    output_wasm: Vec<u8>,
+    /// Information about the build so that it can be reproduced.
+    build_info:  utils::VersionedBuildInfo,
+    /// The sources that were built, archived as a tar file.
+    tar_archive: TarArchiveData,
+}
+
+fn build_in_container<'a>(
+    image: String,
+    package_target_dir: &Path,
+    package_root_path: &Path,
+    extra_args: impl Iterator<Item = &'a String>,
+    container_runtime: &str,
+    out_path: &Path, // canonical, fully expanded path
+    tar_path: &Path, // canonical, fully expanded path
```

Comment:
This should be explained in the docs instead so that you can see it without visiting the function.